src/query/storage/m3/dynamic_cluster.go

```diff
@@ -88,10 +96,397 @@ func NewDynamicClusterOptions() DynamicClusterOptions {
 	}
 }
 
-type dynamicCluster struct{}
+type dynamicCluster struct {
+	clusterCfgs []DynamicClusterNamespaceConfiguration
+	logger      *zap.Logger
+	iOpts       instrument.Options
+
+	sync.RWMutex
+
+	namespaces              ClusterNamespaces
+	unaggregatedNamespace   ClusterNamespace
+	aggregatedNamespaces    map[RetentionResolution]ClusterNamespace
+	namespacesByEtcdCluster map[int]clusterNamespaceLookup
+
+	nsWatches []namespace.NamespaceWatch
+	closed    bool
+}
 
 // NewDynamicClusters creates an implementation of the Clusters interface
 // supports dynamic updating of cluster namespaces.
 func NewDynamicClusters(_ DynamicClusterOptions) (Clusters, error) {
 	return nil, errors.New("dynamic cluster configuration not yet supported")
 }
+
+// TODO(nate): Replace constructor above with this method
+// once namespace staging is complete.
+//
+// newDynamicClusters creates an implementation of the Clusters interface
+// supports dynamic updating of cluster namespaces.
+func newDynamicClusters(opts DynamicClusterOptions) (Clusters, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	cluster := dynamicCluster{
```

Comment:
nit: should this be a `&dynamicCluster`

Reply:
I don't think it really matters since we do `return &cluster` a few lines below.

Reply:
Yeah, just a bit more in keeping with convention